Parse Candidate Extensions (RFC5245)

joeturki opened the pull request:
#### Description
- Rewrote `UnmarshalCandidate` to better align with RFC5245.
- Added new methods for extracting candidate extensions.
- Updated `Equal` and `Marshal` to accommodate these changes.
- Added tests for everything. 
- Fixed a "bug" where tcptype would appear before raddr .
- Made a minor change to gather.go; because we can't use the config struct as a map key anymore.
- Added `CandidateExtension`.

I'm aware this is a breaking change but this change is required for many issues like https://github.com/pion/webrtc/issues/2993

Edit: I'll make it non breaking change tomorrow. 

#### Reference issue

https://github.com/pion/ice/issues/753

Sean-Der commented on candidate_base.go:
```diff
@@ -10,6 +10,7 @@ import (
 	"hash/crc32"
 	"io"
 	"net"
+	"reflect"
```
Just started reading, but would prefer to never use reflection.

Makes things harder on tinygo and we lose safety

Sean-Der commented on candidate.go:
```diff
@@ -52,6 +52,17 @@ type Candidate interface {
+	// In the order of insertion, *(key value).
+	// Extension attributes are defined in RFC 5245, Section 15.1:
+	// https://datatracker.ietf.org/doc/html/rfc5245#section-15.1
+	Extensions() []string
```
What do you think of defining a `Extension` struct? Gives you flexibility to add more fields in the future

joeturki replied:
Done

Sean-Der replied:
Sorry I wouldn’t do `CandidateExtensions‘ as a type. You can do just the slice

Also maybe don’t do a pointer to the slice. A empty slice means no extensions exist is good!

joeturki replied:
@Sean-Der so Extensions() returns a slice of CandidateExtension{key, value} (maybe private?) and GetExtension(key string)?

Sean-Der replied:
I think that makes sense, tell me if you disagree 

* Extensions as a type Allows us to add methods/more fields to Extensions without breaking API
* Returning slice is nice because reading the doc I don’t have to look up  as much. I don’t see a benefit to a type here

joeturki replied:
Okay, so I'm dumb I thought you meant by struct (CandidateExtensions), not a slice of structs, fixing that 👍

Sean-Der commented on candidate_base.go:
```diff
@@ -496,103 +503,382 @@ func (c *candidateBase) Marshal() string {
+		key := c.extensions[i]
+
+		if i+1 >= len(c.extensions) {
+			// should never happen
```
I am on phone, so don’t fully understand code yet. Can/should we return an error in bad situation?

Sean-Der commented:
Nice, love these improvements :)

Glad we can do this without API break. It means people don’t update and new versions don’t get improved :(

joeturki commented on candidate_base.go:
```diff
@@ -406,12 +409,16 @@ func (c *candidateBase) Equal(other Candidate) bool {
 		c.TCPType() == other.TCPType() &&
-		c.RelatedAddress().Equal(other.RelatedAddress())
+		c.RelatedAddress().Equal(other.RelatedAddress()) &&
+		reflect.DeepEqual(c.Extensions(), other.Extensions())
```
I added DeepEqual

joeturki commented on candidate_base.go:
```diff
@@ -496,103 +503,382 @@ func (c *candidateBase) Marshal() string {
+			nil,
+			parserExt.extensions,
+		})
 	default:
```
That was inherited code, fixed thanks :)

joeturki commented:
I added `CandidateExtensions` and `CandidateExtension`, And fixed all the other issues.

@Sean-Der @nils-ohlmeier let me know if there are other problems :)

Thanks.

Sean-Der commented on candidate_base.go:
```diff
@@ -45,6 +45,7 @@ type candidateBase struct {
 
 	remoteCandidateCaches map[AddrPort]Candidate
 	isLocationTracked     bool
+	extensions            *CandidateExtensions
```
I would just do `[]CandidateExtension` does not need to be a pointer

nils-ohlmeier reviewed (approved):
Looks good to me.

Sean-Der commented on candidateextensions.go:
```diff
@@ -0,0 +1,141 @@
+// *(SP extension-att-name SP extension-att-value)
+// Where extension-att-name, and extension-att-value are byte-strings
+// as defined in https://tools.ietf.org/html/rfc5245#section-15.1
+func UnmarshalCandidateExtensions(raw string) (*CandidateExtensions, error) {
```
Do you need to expose?

Sean-Der reviewed (approved):
Fantastic! This is great.

Sorry about the stress of APIs. It’s just so hard to fix them. We can always fix/improve code though.

Its nice to expose as little as possible until someone requests it :)

joeturki commented:
@Sean-Der When you have a time can you take a look.

> Sorry about the stress of APIs. It’s just so hard to fix them. We can always fix/improve code though.

It's not stressing at all for me :) 

> Its nice to expose as little as possible until someone requests it :)

This makes sense, Noted :)

seaduboi-openai commented on candidate.go:
```diff
@@ -52,12 +52,28 @@ type Candidate interface {
+	// Extension attributes are defined in RFC 5245, Section 15.1:
+	// https://datatracker.ietf.org/doc/html/rfc5245#section-15.1
+	//.
+	GetExtension(key string) (value string, ok bool)
```
Mind doing `GetExtension(key string) (value CandidateExtension, ok bool)`

Sean-Der replied:
Sorry this is me, was on work account.

Sean-Der commented:
@joeturki sorry one last change to keep return types consistent. After that merge it! If we do have bugs/improvements to make we call roll forward :)

sorry for all the rounds on this.

joeturki commented:
@Sean-Der 
Hopefully, this is my first API change of many :) 

I now have a better understanding of how changes are made to the Pion API, I'll take all feedback into consideration to try to make future iterations shorter. 

Thanks!
